# Základy programování v Pythonu

V této části se seznámíme s:

* Základy pracování s Jupyter Notebook
* Aritmetickými operacemi pomocí Pythonu
* Řešeními vícestupňových problémů pomocí proměnných
* Vyhodnocováním podmínek pomocí Pythonu
* Kombinacemi podmínek s logickými operátory


## Jak na Jupyter Notebook

#### Co je to notebook?

Notebook integruje kód a jeho výstup do jednoho dokumentu, který kombinuje vizualizace, narativní text, matematické rovnice a další multimédia. Jinými slovy: je to jeden dokument, v němž můžete spustit kód, zobrazit jeho výstup či přidat vysvětlení, vzorce a grafy a zvýšit transparentnost, srozumitelnost a opakovatelnost vaší práce a usnadnit její sdílení.

![New Notebook](images/new-notebook.jpg)

Takto vypadá naše prostředí, do kterého budeme náš kód zapisovat. Skládá se z ovládací palety a buňek, které budeme v průběhu práce pod touto paletou přidávat.

## Buňky

Buňky tvoří tělo notebooku. Na snímku obrazovky nového notebooku výše je prázdnou buňkou pole se zeleným obrysem. Budeme se zabývat dvěma hlavními typy buněk:

* Buňka **kódu** obsahuje kód, který chceme vyhodnotit. Když kód spustíme, notebook zobrazí jeho výstup pod buňkou, která ho vygenerovala. 

* Buňka **Markdown** obsahuje text formátovaný pomocí Markdownu. Při spuštění buňky Markdown se její výstup zobrazí přímo v ní.

První buňka v novém notebooku je vždy buňka kódu.

Pojďme ji společně vyzkoušet s klasickým "Hello World" programem. Vepište do buňky `print("Hello World!")` a poté klikněte na ovládací paletě na tlačítko `run` nebo zmáčkněte klávesovou zkratku `ctrl + enter`.

![Run button](images/run-button.jpg)

In [1]:
print("Hello World!")

Hello World!


Když spustíme buňku, její výstup se zobrazí níže a štítek nalevo se změní z `In []` na `In [1]`. 
Změněné číslo štítku označuje, kdy byla buňka spuštěna – v tomto případě byla buňka spuštěna jako první.

Spusťte buňku znovu a štítek se změní na `In [2]`, protože buňka byla nyní spuštěna jako druhá. Později zjistíme, proč je to užitečné.

Na ovládací paletě klikněte na `Insert` a vyberte `Insert cell below`, abyste pod libovolnou buňkou vytvořili buňku novou, a vyzkoušejte následující kód, abyste zjistili, co se stane. Všimli jste si něčeho jiného?
```
import time
time.sleep(3)

```

In [3]:
import time
time.sleep(3)

Tato buňka neprodukuje žádný výstup, ale trvá tři sekundy, než se vyhodnotí. Všimněte si, jak Jupyter označuje, když je buňka aktuálně spuštěna – změnou jejího štítku na `In [*]`.

## Klávesové zkratky

Možná jste při vyhodnocování svých buněk zpozorovali, že barva jejich ohraničení se změnila na modrou, zatímco při úpravách byla zelená. 
V notebooku Jupyter je vždy jedna „aktivní“ buňka zvýrazněna ohraničením, jehož barva značí režim:
    
* Zelený obrys – buňka je v „režimu úprav“
* Modrý obrys – buňka je v „příkazovém režimu“

Co tedy můžeme dělat s buňkou, když je v příkazovém režimu? Zatím jsme viděli, jak spustit buňku pomocí příkazu `Ctrl + Enter`, ale existuje spousta dalších příkazů, které můžeme použít. Pomocí klávesových zkratek se naučíme využívat Jupyter Notebook mnohem efektivněji.

Klávesové zkratky jsou velmi oblíbeným aspektem prostředí Jupyter, protože usnadňují práci s buňkami. Na aktivní buňce, která je v příkazovém režimu, můžete pomocí klávesových zkratek provádět celou řadu operací.

Níže je uvedený seznam některých klávesových zkratek Jupyteru. Nemusíte si je všechny ihned zapamatovat, tento seznam by vám ale měl poskytnout dobrou představu o tom, co všechno je možné v Jupyteru provádět.

* Mezi režimem úprav a režimem příkazů můžete přepínat pomocí kláves `Esc` a `Enter`.
* Jakmile jste v příkazovém režimu:
 * Posouvejte se nahoru a dolů po buňkách pomocí kláves `Nahoru` a `Dolů`.
 * Stisknutím `A` nebo `B` vložíte novou buňku nad nebo pod aktivní buňku.
 * `M` přepne aktivní buňku z buňky kódu na buňku Markdown.
 * `Y` přepne aktivní buňku na buňku kódu.
 * `D + D` (D dvakrát) vymaže aktivní buňku.
 * `Z` zruší odstranění buněk (vrátí je).
 * Podržte klávesu `Shift` a stiskněte `nahoru` nebo `dolů` a vyberte více buněk najednou. Když je vybráno více buněk, `Shift + M` váš výběr sloučí.

Vyzkoušejte tyto klávesové zkratky ve svém vlastním notebooku. Až budete připraveni, naučíme se společně, jak v našem notebooku formátovat text.

## Markdown

Markdown je lehký a snadno zvládnutelný sázecí jazyk pro formátování prostého textu.


# Tohle je nadpis první úrovně

## Tohle je nadpis druhé úrovně

Toto je prostý text, který tvoří odstavec. Přidejte zvýraznění pomocí **tučné** a __tučné__, nebo *kurzíva* a _kurzíva_. 

Odstavce musí být odděleny prázdným řádkem.

* Někdy chceme zahrnout seznamy.
* Které lze odrážet pomocí hvězdiček. 


1. Seznamy můžou také být číslované
2. Chceme-li seřazený seznam.

[Je možné zahrnout hypertextové odkazy](https://www.example.com)

Vložený kód používá jednoduché obrácené úvozovky: `foo ()` a bloky kódu používají trojité obrácené úvozovky:
```
bar()
``` 
Nebo také může být odsazen čtyřmi mezerami:

    foo()
    
Snadno také můžete přidat obrázky: ![Python Icon](images/Python.png)

## Aritmetické operace s Pythonem

Nejprve si ukážeme, jak použít Python jako kalkulačku. Do buněk notebooku můžeme vepsat matematické výrazy a poté je vyhodnotit. 

Zaměřte se na funkci operátorů `//`, `%`, `**`.

In [4]:
2+3+9

14

In [5]:
2*3+9

15

In [6]:
22.57*-12.36*-3

836.8956

In [7]:
100 / 7

14.285714285714286

In [8]:
100 // 7

14

In [9]:
100 % 7

2

In [10]:
5 ** 3

125

Mužeme vidět, že operátory, jako je `/` a `*`, mají v souladu s matematickými konvencemi přednost před jinými, jako je `+` a `-`. Abychom specifikovali pořadí operací ve výrazu, můžeme klasicky použít závorky.

In [ ]:
((2 + 5) * (17 - 3)) / (4 ** 3)

V základu podporuje Python tyto matematické operátory:

| Operátor   | Funkce           | Příklad     | Výsledek    |
|------------|-------------------|-------------|-----------|
| `+`        | Sčítání          | `2 + 3`     | `5`       |
| `-`        | Odčítání       | `3 - 2`     | `1`       |
| `*`        | Násobení    | `8 * 12`    | `96`      |
| `/`        | Dělení          | `100 / 7`   | `14.28..` |
| `//`       | Celočíselné dělení    | `100 // 7`  | `14`      |    
| `%`        | Zbytek dělení | `100 % 7`   | `2`       |
| `**`       | Exponent          | `5 ** 3`    | `125`     |

Pokuste se vyřešit několik následujícíh úloh. Využijte k tomu následující prázdné buňky nebo si přidejte další navíc.

1) Populace města je 198 568. Z toho 45 312 jsou muži a 35 678 jsou ženy. Kolik je v tomto městě dětí?

2) Cena 21 TV je 652 121 korun. Kolik stojí jedna TV?

3) Továrna na žárovky vyrobí 12 652 žárovek za měsíc. Kolik jich vyrobí za rok?

Tyto úlohy se pochopitelně hodí spíše pro žáky základní školy, ale zde nám vystačí jako příklad.

## Řešení vícestupňových problémů pomocí proměnných

Pokusíme se vyřešit následující slovní úlohu pomocí Pythonu. 

> Obchod prodává balíky s ledem za 20 korun a vydělá na nich 20 % jejich ceny. Jaký bude zisk obchodu, pokud prodá 500 balíků?

Úlohu můžeme rozepsat do formy, se kterou už si Python poradí.

*Cena balíku* = 20

*Marže* = 20 % = .2

*Výdělek na balíku* = marže * cena balíku = 0.2 * 20

*Počet prodaných balíků* = 500

*Zisk* = počet prodaných balíků * výdělek na balíku = 500 * (.2 * 20)

In [11]:
500*(0.2*20)

2000.0

Obchod si tak přijde na 2000 korun. Úlohu takto sice řešit můžeme, ale není to úplně nejpřehlednější způsob. Na první pohled totiž nevidíme, co čísla v buňce znamenají. Číslům můžeme přidělit jména tak, že vytvoříme Python *proměnné*.


> **Proměnné**: Pří práci s programovacími jazyky, jako je Python, ukládáme informace do *proměnných*. Proměnné si můžete představit jako krabičky na ukládání dat. Data, která jsou uložena v proměnné, nazýváme hodnotami.

In [12]:
cena_balíku = 20

In [13]:
marže = 0.2

In [14]:
počet_prodaných_balíků = 500

Proměnné `cena_balíku`, `marže`, a `počet_prodaných_balíků` teď obsahují informace zadané ve slovní úloze. Hodnotu proměnné si můžeme zkontrolovat tak, že ji napíšeme do buňky. Proměnné můžeme kombinovat pomocí matematických operátorů.

> **Doplňování kódu**: Při psaní jmen již vzniklých proměnných nám Jupyter Notebook pomáhá a po napsání prvních písmen můžeme zmáčknout na klávesnici `Tab` pro automatické doplnění zbytku.

In [ ]:
počet_prodaných_balíků

Pokud zkusíme napsat proměnnou, která ještě nebyla *definována*, Python nám vyhodí error.

In [ ]:
zisk

In [15]:
zisk = počet_prodaných_balíků*(marže*cena_balíku)

In [16]:
zisk

2000.0

Ukládání dat a jejich manipulace za pomocí vhodně pojmenovaných proměnných je skvělý způsob, jak jednoduše vysvětlit, co váš kód dělá.


Zobrazme si výsledek slovní úlohy pomocí jednoduché zprávy. Můžeme pro to použít Python *funkci* `print`.

> **Funkce** je opakovaně použitelná sada pokynů. Vyžaduje jeden nebo více vstupů, provádí určité operace a často vrací výstup. Python poskytuje mnoho vestavěných funkcí, jako je `print`, a také nám umožňuje definovat si funkce vlastní. 

In [17]:
print("Obchod má zisk", zisk)

Obchod má zisk 2000.0


> **`print`** funkce je využívána pro zobrazení informací. Má jeden nebo více vstupů, které mohou být například ve formě textu (mezi uvozovkami, např., `"Tohle je text!"`) nebo čísel. Více se o proměnných a funkcích dozvíme později.

Vytvoření buněk pro každou proměnnou nebo matematickou operaci je nepraktické. Naštěstí nám Jupyter notebook dovoluje psát mnoho řádků kódu do samostatné buňky. Výsledek posledního řádku kódu je potom zobrazen jako výstup. 

Přepíšeme tedy celou slovní úlohu do jediné buňky.


In [18]:
# Vstupní data
cena_balíku = 20
marže = 0.2
počet_prodaných_balíků = 500

# Provedení kalkulací
zisk_na_balik = cena_balíku * marže
zisk = počet_prodaných_balíků * zisk_na_balik

# Zobrazení výsledku
print("Obchod má zisk", zisk)

Obchod má zisk 2000.0


Můžete si všimnout, že používáme `#`, abychom označili komentáře k našemu kódu.

> **Komentáře** a prázdné řádky jsou během vyhodnocování kódu ignorovány, pro nás jsou ale užitečné, protože nám předávají informace o tom, co vlastně kód dělá.

Jednořádkové komentáře začínájí znakem `#`, zatímco víceřádkové komentáře jsou oddělené trojitými uvozovkami `"""`. Uveďme si několik příkladů.

In [19]:
super_cislo = 2 # Tohle číslo je super

In [20]:
záporné_cislo = -2
# Tohle číslo je záporné

In [21]:
"""
Toto je víceřádkový komentář, do kterého toho můžete napsat, kolik chcete.

Komentáře jsou velmi užitečné pro čitelný kód, vyhněte se ale dlouhému rozepisování, které dělají kód nepřehlednější.

Pokud používáte smysluplná jména proměnných, kód je většinou srozumitelný.

"""
nula = 0

## Vyhodnocování podmínek v Pythonu

Kromě aritmetických operací nám Python nabízí i několik relačních operací pro porovnání čísel nebo proměnných.

| Operátor    | Popis                                                     |
|-------------|--------------------------------------------------------------  ---|
| `==`        | Zkontroluje, zda jsou si operandy rovny                           |
| `!=`        | Zkontroluje, zda si operandy rovny nejsou                         |
| `>`         | Zkontroluje, zda operand vlevo je větší než operand vpravo         |
| `<`         | Zkontroluje, zda operand vlevo je menší než operand vpravo         |
| `>=`        | Zkontroluje, zda operand vlevo je větší nebo roven operandu vpravo |
| `<=`        | Zkontroluje, zda operand vlevo je menší nebo roven operandu vpravo |

Výsledek relačních operací je buď pravda `True`, nebo lež `False` (Všimněte si velkých písmen u True a False). Toto jsou speciální klíčová slova pro Python. Vyzkoušíme si pár příkladů.

In [22]:
první_číslo = 2
druhé_číslo = 5
třetí_číslo = 9

In [23]:
první_číslo == 2 #Zkouška rovnosti

True

In [24]:
první_číslo == druhé_číslo #Zkouška rovnosti

False

In [25]:
první_číslo != druhé_číslo #Zkouška nerovnosti

True

In [26]:
druhé_číslo != 5 #Zkouška nerovnosti

False

Tak jako můžeme do proměnných uložit výsledek operací aritmetických, můžeme do proměnných uložit i výsledek operací relačních.

In [27]:
cena_balíku = 20
je_balík_ledu_drahý = cena_balíku >= 30
print("Je balík ledu drahý?", je_balík_ledu_drahý)

Je balík ledu drahý? False


## Kombinace podmínek s logickými operátory

Logické operátory `and`, `or` a `not` operují na podmínkách a také na `True` a `False` hodnotách (Také označovány jako *boolean* hodnoty). 

Operátor `and` nám vrácí `True`, pouze pokud obě podmínky jsou `True`. 

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |

In [ ]:
první_číslo

In [ ]:
první_číslo > 0 and první_číslo <= 3

In [ ]:
True and False

In [ ]:
True and True

In [ ]:
False and False

Operátor `or` nám vrací `True`, pokud alespoň jedna z podmínek je `True`. `False` nám vrátí tehdy, pokud jsou obě podmínky `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |


In [ ]:
druhé_číslo

In [ ]:
druhé_číslo == 5 or druhé_číslo < 3

In [ ]:
druhé_číslo != 5 or druhé_číslo < 3

In [ ]:
False or False

Operátor `not` nám vrací `False` pokud je podmínka `True` a vrací nám `True` pokud je podmínka `False`.

In [ ]:
druhé_číslo

In [ ]:
not druhé_číslo == 5

In [ ]:
not True

In [ ]:
not True and False

Logické operátory můžeme kombinovat tak, aby tvořily komplexní podmínky. Pomocí kulatých závorek udáváme pořadí tak jako u matematických operací.

Pokud žádné závorky nepoužijeme, výraz se vyhodnotí zleva doprava.

In [ ]:
not (True and False)

## Cvičení

#### Slovní úlohy
1. Autobus má průměrnou spotřebu 10 l/100 km a jede z bodu A do bodu B, které jsou od sebe vzdálené 160 km. Kolik paliva autobus spotřebuje?
2. Prodavač na trhu zeleniny prodává okurky za 20 Kč/kg a rajčata za 35 Kč/kg. Za týden prodal 35 kg okurek a 30 kg rajčat, ale v průběhu týdne musel snížit cenu okurek na 18 Kč/kg, protože se objevil konkurenční stánek. V tu chvíli měl prodaných 20 kg okurek. Kolik byl jeho celkový zisk za týden?

#### Podmínky a logické operátory
Vymyslete si tři čísla a proveďte na nich tři příklady porovnávacích operací, tři příklady s logickými operátory a dva příklady u kterých zkombinujete oba druhy operátorů.